### ISSUE TITLE
Make units that have the same stats to have the same checksum as in 1.16

### ISSUE THREAD

**Issue by demario-w** [wesnoth/wesnoth#7188]
### Describe the desired feature

To replay saves from 1.16 smoothly, it is required that units have the same checksum if the stats and traits have not changed.

The units with a weapon special in 1.17 (horseman, thief, shaman, mage...) have different checksums than what they did in 1.16. Example for an intelligent, quick horseman:
```
(1.16) checksum="B0xNnBFOKh>Lw-f-d9Jcec.F<6D.>568p3fcgwbBc5kj.b25Y7y2Ohe5l+84Alnp.gm<39sh139,+n<pzlu8dYUjp0kF4mcn9V7fD>4I3bs7l3jv>0IuvQDBp8L+M62,"
(1.17) checksum="BllDfGwgr>>yfbf,f4Jk1naz.a.Xhlc8,3cebP4dp4>n2h><e0F-Wcg+8jlbZ+jk.oh<,mB6IZ++llbW,qhiZYwaIX<7iuDVF<.yfmF>5F3M3u59s-sujKDdp5+ijg1K"
```

If the horseman loses charge special, the checksums are the same in both version. This is most likely due to the deprecation of SPECIAL_NOTES. When loading the save, the game is throwing the warning:

_error deprecation: SPECIAL_NOTES has been deprecated and removed.; Use [special_note]note= tags instead to apply special notes, or use the {NOTE_*} macros (note the singular)._

**Comment by Pentarctagon** [comment]
Sounds like a similar sort of issue as #5743

**Comment by demario-w** [comment]
When loading the 1.16 save in both version with --log-debug=unit, the logging in src/units/unit.cpp[get_checksum] gives different results:
```diff
-- int.qui.horseman_checksum-1.16.out  2022-12-03 22:41:17.326940154 +0800
+++ int.qui.horseman_checksum-1.17.out  2022-12-03 22:42:32.711605434 +0800
@@ -28,7 +28,6 @@
            apply_to = both
            id = charge
            multiply = 2
-           name = charge
            special_note = Using a charging attack doubles both damage dealt and received; this does not affect defensive retaliation.
        [/damage]
    [/specials]
```
The absence of the 'name' attribute (inside attack/specials/damage) in 1.17 is what makes the checksum to fail. If the following change is made to the version 1.17:
```diff
diff -r 4133771678ff src/units/unit.cpp
--- a/src/units/unit.cpp	Fri Dec 02 10:04:35 2022 +0800
+++ b/src/units/unit.cpp	Sat Dec 03 23:06:57 2022 +0800
@@ -2739,7 +2739,6 @@ std::string get_checksum(const unit& u)
 
 			child_spec.recursive_clear_value("description");
 			child_spec.recursive_clear_value("description_inactive");
-			child_spec.recursive_clear_value("name");
 			child_spec.recursive_clear_value("name_inactive");
 		}
 	}
```
then the name is present in the unit that gets checksum-ed and the test pass.
Provided that the unit WML is the same and the code for get_checksum are identical for both versions, that must come from a change inside the underlying recursive_clear_value() that generates a different output.
It is fair to think that the 4 calls to recursive_clear_value() used to do nothing in 1.16, it got "fixed" in 1.17 and now the checksum fails.

**Comment by Pentarctagon** [comment]
That fix for adding those attributes to be ignored is, unfortunately, valid and necessary. So it's not something that can be reverted.

**Comment by demario-w** [comment]
> That fix for adding those attributes to be ignored is, unfortunately, valid and necessary.

Interesting. Can you point to the discussion about that fix and how was reached the conclusion that it was an urgent problem to solve? I would like to understand better the whole picture. Thanks.

**Comment by Pentarctagon** [comment]
It's the issue I linked above.

**Comment by demario-w** [comment]
> It's the issue I linked above.

I see. I still have a hard time understanding how a code that is equally present in branches 1.16 and 1.17 could be the cause of different computed checksum between these branches. That is why I was expecting a reference to another change.

If that issue #5743 is containing all the discussion on the topic, I think it's hardly a convincing analysis of the problem or the solution. The fact that BFW 1.16 is still generating checksum that takes the name attribute into account shows that this "fix" is not solving anything.

> That fix is valid

The only question about testing of that fix (in the pull-request) has been answered [with](https://github.com/wesnoth/wesnoth/pull/6106#issuecomment-927196053) "I was unable to test". So hardly a proof that is valid. There is also no decision that acknowledge that this is breaking checksum with past versions and that it is something that is acceptable.

> ... and necessary

The original problem is the use of an undocumented mechanism to [implement](https://github.com/wesnoth/wesnoth/issues/5743#issue-869085758) someone's "macro-based translation" (in an add-on, I assume). So this necessity of breaking compatibility of the engine behavior to accommodate the needs from an add-on author is quite unusual.

Of course it make sense that description are not part of the data used to compute the checksum (maybe that's what you mean by "valid"). But given that it has been the case since years, I think it makes even less sense to break compatibility to fix a problem no one has noticed until recently.

**Comment by Pentarctagon** [comment]
> I see. I still have a hard time understanding how a code that is equally present in branches 1.16 and 1.17 could be the cause of different computed checksum between these branches. That is why I was expecting a reference to another change.
> 
> If that issue https://github.com/wesnoth/wesnoth/issues/5743 is containing all the discussion on the topic, I think it's hardly a convincing analysis of the problem or the solution. The fact that BFW 1.16 is still generating checksum that takes the name attribute into account shows that this "fix" is not solving anything.

It is clearly not present in 1.16: https://github.com/wesnoth/wesnoth/blob/1.16/src/units/unit.cpp#L2703-L2707
In 1.16 and earlier it cleared the description, description_inactive, name, and name_inactive attributes from being considered in the checksum for abilities and traits but only the description attribute for weapon specials. In 1.17+ it now ignores the same four attributes for weapon specials as well.

**Comment by Pentarctagon** [comment]
If @ForestDragon-wesnoth is around, he can comment on the actual issue their macro-based translation ran into in more depth.

**Comment by demario-w** [comment]
> > I think it's hardly a convincing analysis of the problem or the solution. The fact that BFW 1.16 is still generating checksum that takes the name attribute into account shows that this "fix" is not solving anything.
> 
> It is clearly not present in 1.16: https://github.com/wesnoth/wesnoth/blob/1.16/src/units/unit.cpp#L2703-L2707 In 1.16 and earlier it cleared the description, description_inactive, name, and name_inactive attributes from being considered in the checksum for abilities and traits but only the description attribute for weapon specials. In 1.17+ it now ignores the same four attributes for weapon specials as well.

I have posted a diff on the dump of wcfg on line 2743 that shows that the name attribute is not removed in 1.16. The code is there in 1.16 but it doesn't work as intended. Hence the break in behavior in 1.17 (where something [else] has changed?)
 
> > The original problem is the use of an undocumented mechanism to [#5743 (comment)](https://github.com/wesnoth/wesnoth/issues/5743#issue-869085758) someone's "macro-based translation" (in an add-on, I assume). So this necessity of breaking compatibility of the engine behavior to accommodate the needs from an add-on author is quite unusual.
> 
> Not really. Many developers are themselves add-on authors, and make changes to the engine to allow them to do certain things in their add-ons that weren't possible before.

The "need" for a developer to break compatibility in mainline for the interest of his own pet-project is kinda… how should I put it… it's a bit [entitled](https://github.com/wesnoth/wesnoth/issues/6971#issuecomment-1221186649)

**Comment by Pentarctagon** [comment]
> I have posted a diff on the dump of wcfg on line 2743 that shows that the name attribute is not removed in 1.16. The code is there in 1.16 but it doesn't work as intended. Hence the break in behavior in 1.17 (where something [else] has changed?)

I don't know what wcfg is, but I linked previously to the current 1.16 code which doesn't have those attributes being removed. If you think it's actually present, then please link to that code on github so I can be sure of what you're referring to.

**Comment by Discontinuum** [comment]
> I don't know what wcfg is,

I think he means that
https://github.com/wesnoth/wesnoth/blob/e7903e4eac18e086cb7f8203aba3af719813d5cf/src/units/unit.cpp#L2743

**Comment by Pentarctagon** [comment]
Ah, alright. For there as well though, the name attribute is not being removed, only description: https://github.com/wesnoth/wesnoth/blob/e7903e4eac18e086cb7f8203aba3af719813d5cf/src/units/unit.cpp#L2696-L2708

So I'm still unsure where the code that's removing that attribute would be in 1.16.

**Comment by stevecotton** [comment]
I understand both sides of this, but #7200 seems a simple solution to both.

**Comment by soliton-** [comment]
It's a little mind boggling to me how hard it seems to understand that there were attributes included in the checksum that make no sense to be included and this was fixed. No developer entitlement needed at all. There is no expectation that 1.16 is compatible to 1.18 the same as for other stable versions since the entirety of wesnoth's development.

If you want to suggest some way to improve compatibility then by all means do that. I think we're at a point where something like what stevecotton proposes above can be reasonable. A request to keep unit stats or the checksum calculation the same however is ridiculous.

**Comment by stevecotton** [comment]
Closing as #7200 is merged, so 1.17 will now accept the old checksums too.
